### PR TITLE
X-ray windows transmission. Metadata class added

### PR DESCRIPTION
**Pull request by jgalan**
Large jgalan  [](https://gitlab.cern.ch/rest-for-physics/axionlib/-/commits/windows)

- `TRestAxionXrayWindow` metadata class added.
    - Validation pipeline added
    - Documentation added
    - Test image generated

Example definition added:
```
    <TRestAxionXrayWindow name="siliconFoil" verboseLevel="info">
        <parameter name="center" value="(0,0,0)mm" />

        <parameter name="type" value="foil" />
        <parameter name="thickness" value="0.3um" />
        <parameter name="material" value="Si3N4" />
        <parameter name="radius" value="8mm" />
    </TRestAxionXrayWindow>

    <TRestAxionXrayWindow name="strongBack" verboseLevel="info">
        <parameter name="center" value="(0,0,0)mm" />

        <parameter name="windowType" value="grid" />
        <parameter name="thickness" value="200um" />
        <parameter name="material" value="Si" />
        <parameter name="radius" value="8mm" />

        <parameter name="patternWidth" value="0.5mm" />
        <parameter name="patternGap" value="4mm" />
        <parameter name="patternOffset" value="2mm" />
    </TRestAxionXrayWindow>

    <TRestAxionXrayWindow name="cathode" verboseLevel="info">
        <parameter name="center" value="(0,0,0)mm" />

        <parameter name="type" value="foil" />
        <parameter name="thickness" value="0.02um" />
        <parameter name="material" value="Al" />
        <parameter name="radius" value="8mm" />
    </TRestAxionXrayWindow>
```

windowsTest

### CODE REVIEW

**Comment by jgalan** [comment]
This PR is ready for review and it could be merged to master

**Comment by jovoy** [comment]
Hey! I just remembered that back in the day we built a method to calculate the window strip width and distance for any window given the radius, number of strips and open aperture. I could implement it later but it could be good for trying out windows because the companies give you the open aperture that is possible and then you might change the number of strips and this way you won't have to calculate it by hand:

proc calcWindowVals(radiusWindow: MilliMeter,
                    numberOfStrips: int,
                    openApertureRatio: float): tuple[width: MilliMeter,
                                                      dist: MilliMeter] =
  let
    totalArea = π * radiusWindow * radiusWindow
    areaOfStrips = totalArea * (1.0 - openApertureRatio)
    #width and distance between strips per strip;
    #the width on both sides is a whole width
    #(Don't know what to do about the additional string width)
    #not important at high strip number but at low like CAST
    dAndwPerStrip = radiusWindow * 2.0 / (numberOfStrips.float + 1.0)
  var
    lengthStrip: mm
    lengthAllStrips: mm

  for i in 0..(numberOfStrips/2).round.int - 1:
    lengthStrip = sqrt(radiusWindow * radiusWindow -
      (i.float * dAndwPerStrip + 0.5 * dAndwPerStrip) *
        (i.float * dAndwPerStrip + 0.5 * dAndwPerStrip)
    ) * 2.0
    lengthAllStrips = lengthAllStrips + lengthStrip
    echo lengthStrip

  lengthAllStrips = lengthAllStrips * 2.0

  let
    widthStrips = areaOfStrips / lengthAllStrips
    distStrips = dAndwPerStrip - widthStrips
  echo widthStrips
  echo distStrips
  result = (width: widthStrips, dist: distStrips)

**Comment by jgalan** [comment]
Hi @jovoy probably that comment would be better placed at a new issue?

The present PR is for a REST metadata class that allows to create a window once you know all the parameters of the window.

I understand what you propose is to obtain those parameters from your method. This could be implemented here, or one could have a python script that generates the RML definition directly. That would be an option.

Still, I believe the best is to create an issue to have it in mind, and discuss about how to implement it.

**Comment by jgalan** [comment]
Perhaps @jovoy @Vindaar could approve this PR and create the issue?

File: src/TRestAxionXrayWindow.cxx [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+    FILE* fin = fopen(materialFileName.c_str(), "rt");
+
+    fEnergy.clear();
+    fTransmission.clear();
+    double en, value;
+    while (fscanf(fin, "%lf\t%lf\n", &en, &value) != EOF) {
+        debug << "Energy : " << en << "eV -- Abs : " << value << endl;
+
+        fEnergy.push_back(en / 1000.);
+        fTransmission.push_back(TMath::Power(value, fThickness * units("um")));
```
Comment (Vindaar) [inline src/TRestAxionXrayWindow.cxx:189]: For what thickness are the stored data files? It's unclear from the file names. Just hard to reason whether the transmission calculation for a specific thickness makes sense that way.

File: src/TRestAxionXrayWindow.cxx [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+    }
+
+    return true;
+}
+
+///////////////////////////////////////////////
+/// \brief It returns the vector element index, from `fEnergy`, that is just below the given input energy.
+///
+Int_t TRestAxionXrayWindow::GetEnergyIndex(Double_t energy) {
+    for (unsigned int n = 0; n < fEnergy.size(); n++)
```
Comment (Vindaar) [inline src/TRestAxionXrayWindow.cxx:283]: Avoid `unsigned int` if you need to do subtractions with them. You don't want an underflow to happen. Also means you get a "nice" implicit type cast from `n` ⇒ whatever the compiler deems `n - 1` to be ⇒ `Int_t`.

File: src/TRestAxionXrayWindow.cxx [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+
+    Double_t energyIndex = GetEnergyIndex(energy);
+
+    // Transmission
+    double y2 = fTransmission[energyIndex + 1];
+    double y1 = fTransmission[energyIndex];
+
+    // Normalized field
+    double x2 = fEnergy[energyIndex + 1];
+    double x1 = fEnergy[energyIndex];
```
Comment (Vindaar) [inline src/TRestAxionXrayWindow.cxx:220]: `GetEnergyIndex` returns `-1` for energies outside the range. That leads to hard to debug out of bounds access errors for bad inputs in the usage here.

File: src/TRestAxionXrayWindow.cxx [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+    double n = y1 - m * x1;
+
+    if (m * energy + n < 0) {
+        ferr << "TRestAxionXrayWindow::GetAbsorptionCoefficient. Negative coefficient!" << endl;
+        cout << "y2 : " << y2 << " y1 : " << y1 << endl;
+        cout << "x2 : " << x2 << " x1 : " << x1 << endl;
+        cout << "m : " << m << " n : " << n << endl;
+        cout << "E : " << energy << " bin : " << energyIndex << endl;
+        GetChar();
+    }
```
Comment (Vindaar) [inline src/TRestAxionXrayWindow.cxx:232]: Is `GetChar` similar to the normal C `getchar`? If so what is this block supposed to be for aside from a debugging help and stopping execution in case this branch happens?

File: src/TRestAxionXrayWindow.cxx [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+
+    return (m * energy + n);
+}
+
+///////////////////////////////////////////////
+/// \brief It returns true if the window pattern is hitted. False otherwise.
+///
+Bool_t TRestAxionXrayWindow::HitsPattern(Double_t x, Double_t y) {
+    if (fWindowType == "stripped") {
```
Comment (Vindaar) [inline src/TRestAxionXrayWindow.cxx:241]: If `windowType` was an `enum` this could be a switch statement instead of the string comparisons.

File: src/TRestAxionXrayWindow.cxx [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+        Double_t xEval = fPatternWidth / 2. + x - fPatternOffset;
+
+        if (xEval > 0) {
+            while (xEval > fPatternGap) xEval -= fPatternGap;
```
Comment (Vindaar) [inline src/TRestAxionXrayWindow.cxx:245]: Do we verify somewhere that `fPatternGap` is never `0`? Otherwise this runs into an endless loop.

Same in the other `while` loops below.

edit: `0` is even the default value for `fPatternGap`!

**Review by Vindaar — CHANGES_REQUESTED** [review]
Most comments are minor, but some of them should be fixed.

File: pipeline/metadata/transmission/windowPlot.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+totalSamples = 100000
+
+cathode = ROOT.TRestAxionXrayWindow("windows.rml", "cathode")
+strongBack = ROOT.TRestAxionXrayWindow("windows.rml", "strongBack")
+siFoil = ROOT.TRestAxionXrayWindow("windows.rml", "siliconFoil")
+
+radius = cathode.GetWindowRadius()
+print ( "\nGetting window radius" )
+if( radius != 8 ):
```
[the remark this answers is not in the document]
Comment (jgalan) [reply pipeline/metadata/transmission/windowPlot.py:35]: So, you propose to change by?

File: pipeline/metadata/transmission/windowPlot.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+#    gr.GetXaxis().SetTitleSize(0.05);
+#    gr.GetXaxis().SetLabelSize(0.05);
+#    gr.GetYaxis().SetTitle("Y [mm]")
+#    gr.GetYaxis().SetTitleOffset(1)
+#    gr.GetYaxis().SetTitleSize(0.05);
+#    gr.GetYaxis().SetLabelSize(0.05);
+#    graphsOP.append(gr)
+
+
+rnd = TRandom3(0)
```
[the remark this answers is not in the document]
Comment (jgalan) [reply pipeline/metadata/transmission/windowPlot.py:61]: You are right. We do fix the seed value on event processing that needs to produce always the same results for validation pourposes. But in this case I believe is indifferent.

File: pipeline/metadata/transmission/windowPlot.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+#for n in range(1,rings):
+#    graphsOP[n].Draw("P")
+
+c1.Print(outfname)
+
+print ("All tests passed!  [\033[92m OK \x1b[0m]")
+
+print ("")
+
+exit(0)
```
[the remark this answers is not in the document]
Comment (jgalan) [reply pipeline/metadata/transmission/windowPlot.py:148]: For me is better, so it is very clear it returns 0

File: pipeline/metadata/transmission/windowPlot.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Vindaar) [reply pipeline/metadata/transmission/windowPlot.py:35]:
```suggestion
if radius != 8:
```
is the normal way to write it in python

File: src/TRestAxionXrayWindow.cxx [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+/// We may define the window type using:
+///
+/// - **type**: We may define different options that define the regions
+/// of space where the window material will be effective.
+///   -# **foil**: It defines a homogeneous region of material, all the
+///    region inside the window radius will present the same efficiency.
+///   -# **stripped**: It defines a stripped pattern where the absorption
+///   will be effective, otherwise, if the photon is inside the window
+///   but it doesnt hit the stripped pattern, the transmission will be
+///   equal to 1.
```
[the remark this answers is not in the document]
Comment (jgalan) [reply src/TRestAxionXrayWindow.cxx:53]: In reality a real x-ray window is built using a combination of `TRestXrayWindow`. So, when you define the type *grid* or the type *strip* if the pattern is hitted then we apply the transmission curve using the material the pattern is made of. But if you do not hit the pattern, then the photon just goes through. There is vacuum.

If you build the window as a combination of Si grid + Al foil, then you are combining two `TRestXrayWindow` descriptions. The second one will apply the Al transmission everywhere, since it is a foil type (as soon as the particle falls inside the window).

>> Transmission is 1 if it doesn't hit the pattern? That seems weird.

So, yes, if it doesn't hit the pattern the photon just goes through.

>> Hitting the strips should have a very low (but not 0) transmission. 

Right, the window is built with a particular material and thickness.

>> Hitting the non strip parts should use the window thickness, so transmission somewhere between 0 and 1.

As I tried to explain this is the combination of 2 `TRestAxionXrayWindow` components. You could perfectly define a grid with vacuum in between strips, of course we do not want that, but in principle one could. You need to add an Al foil so that it is not vacuum.

Then, the idea is that inside the process one will select a combination of structures and/or foils to create a more complex window.

File: src/TRestAxionXrayWindow.cxx (hunk 1 above)
Comment (jgalan) [reply src/TRestAxionXrayWindow.cxx:189]: The transmission data (in the above code named `value`) is expressed in `um`. The fact that is expressed in `um` is described at the README.md in the data base.

https://github.com/rest-for-physics/axionlib-data/tree/master/transmission

I think it is important that the coding is correct here. But I believe this will be a rather static code that reads the file. If we add new materials those must be added for `1um`, and someone uploading new materials should be careful enough to read the README.md from the database.

So, basically, we get the transmission for 1um, for 2um you just need to square the transmission, right? Thats what the formula at line 189 is doing.

`fThickness` is expressed in `mm` because is the default units in REST, so thats why `units("um")` so that it is expressed finally in `us`.

File: src/TRestAxionXrayWindow.cxx (hunk 3 above)
Comment (jgalan) [reply src/TRestAxionXrayWindow.cxx]: Fixed at d30f63f8

File: src/TRestAxionXrayWindow.cxx (hunk 5 above)
Comment (jgalan) [reply src/TRestAxionXrayWindow.cxx:246]: Aham

File: src/TRestAxionXrayWindow.cxx (hunk 6 above)
Comment (jgalan) [reply src/TRestAxionXrayWindow.cxx:257]: Should be solved at f45eedcdbf52d348dc3dae7e9f6bff20f5527fe3

File: src/TRestAxionXrayWindow.cxx (hunk 2 above)
Comment (jgalan) [reply src/TRestAxionXrayWindow.cxx]: Fixed at f488660b4

File: inc/TRestAxionXrayWindow.h [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+#include <TRestMetadata.h>
+
+//! A metadata class to create x-ray transmission window definitions
+class TRestAxionXrayWindow : public TRestMetadata {
+   private:
+    /// Position of the center of the window
+    TVector3 fCenter = TVector3(0, 0, 0);  //<
+
+    /// Type of window (foil, stripped, ...)
```
[the remark this answers is not in the document]
Comment (jgalan) [reply inc/TRestAxionXrayWindow.h]: Done.

This is important, it is a ROOT thing. 

If data members are specified using `//<` the value will be written to disk inside the ROOT file.

If data members are specified using `//!` they will be ignored in the write to file. Basically those data members are initialised when the class is being used, or can be used as auxilar variables.

[the remark this answers is not in the document]
Comment (jgalan) [reply inc/TRestAxionXrayWindow.h]: Comment issue was addressed at 8224304ef79817d8a34c25bf942454bba749b610

File: inc/TRestAxionXrayWindow.h [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+    std::string fWindowType = "foil";  //<
+
+    /// Thicknesss of window material
```
[the remark this answers is not in the document]
Comment (jgalan) [reply inc/TRestAxionXrayWindow.h]: All parameters use standard REST units, `mm`, `keV`, `us`.

Changes were applied at 8224304ef79817d8a34c25bf942454bba749b610

File: pipeline/metadata/transmission/windowPlot.py [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+
+ROOT.gSystem.Load("libRestFramework.so")
+ROOT.gSystem.Load("libRestAxion.so")
+
+### Creating a canvas and pad for drawing
+c1 = TCanvas( 'c1', 'My canvas', 1200,400 )
+c1.GetFrame().SetBorderSize( 6 )
+c1.GetFrame().SetBorderMode( -1 )
+
+pad1 = TPad("pad1","This is pad1",0.01,0.02,0.99,0.97);
```
[the remark this answers is not in the document]
Comment (jgalan) [reply pipeline/metadata/transmission/windowPlot.py]: Fixed at 002e962fa85a

File: inc/TRestAxionXrayWindow.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (Vindaar) [reply inc/TRestAxionXrayWindow.h]: > If data members are specified using //< the value will be written to disk inside the ROOT file.
> If data members are specified using //! they will be ignored in the write to file. Basically those data members are initialised when the class is being used, or can be used as auxilar variables.

Woah, that sounds horrific. :rofl:

File: pipeline/metadata/transmission/windowPlot.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Vindaar) [reply pipeline/metadata/transmission/windowPlot.py:61]: Even if in this case we don't currently use it as a direct validation, I still think a reproducible result is preferable.

[the remark this answers is not in the document]
Comment (jgalan) [reply pipeline/metadata/transmission/windowPlot.py:61]: I would prefer to calculate the radius average for the points falling in a particular ring and check the radius is at the appropiate ring rather than comparing pngs.

That would be also independent from `TRandom` generation, and from any `png` library or `ROOT` library update that, let's imagine, changes the color of the pixel, or the default font, or whatever else which is irrelevant to our routine.

[the remark this answers is not in the document]
Comment (jgalan) [reply pipeline/metadata/transmission/windowPlot.py:61]: This validation for the moment just checks the plot is really generated (i.e. the script gets to the end without problems). Further tests may be added in the future.

[the remark this answers is not in the document]
Comment (Vindaar) [reply pipeline/metadata/transmission/windowPlot.py:61]: > I would prefer to calculate the radius average for the points falling in a particular ring and check the radius is at the appropriate ring rather than comparing png images.

I'm not proposing to compare PNG images (even though I do it [here](https://github.com/Vindaar/ggplotnim/blob/master/tests/tCompareRecipes.nim#L59) myself :hand_over_mouth: ). Your proposal sounds like a decent validation.

But I'm saying that a validation pipeline shouldn't introduce variability where not required. Simply choose a fixed seed. In the worst case you never notice it. In the best case it avoids you or someone else at some point debugging why something changes each run (e.g. someone extends some pipeline etc.). Also the `0` is kind of a magic number, that *appears* to be a fixed seed when it isn't. That's misleading in addition to that.

And still, given that the plot is committed, it will lead to each run causing the plot to appear in unstaged changes, which is simply annoying.

[the remark this answers is not in the document]
Comment (jgalan) [reply pipeline/metadata/transmission/windowPlot.py:61]: Then, I reply you with my first answer yesterday and we enter in an infinite loop.

**Comment by jgalan** [comment]
I don't see big problems to merge this PR to master. The only reason to block a PR from merging is that there is no validation, no proper documentation, there are errors or major concerns affecting other parts of the code. But I dont think it is the case at this PR.

If there are further contributions envisaged, those could be placed at an issue for future implementation.

Plz, approve it.

**Review by jovoy — APPROVED** [review]
